proxy/proxy/client.go

```diff
@@ -225,6 +228,15 @@ func (c *Client) Dial(instance string) (net.Conn, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	// Refresh cert 5 minutes before it expires.
+	timer := time.NewTimer(cfg.Certificates[0].Leaf.NotAfter.Sub(time.Now()) - refreshCertBuffer)
+	go func() {
```

Comment:
I think this makes more to put in refreshCfg since we don't want to start more than one goroutine at a time for refresh (and refreshCfg is the one with the mutex).

Reply:
In refreshCfg it would still look something like:
go func() {
    <- time.After(blah)
    c.refreshCfg(instance)
}() (in addition to the mutex changes mentioned in the other comment) Otherwise I'm not sure what you mean. Could you elaborate?

Reply:
Yep, just like that. The main difference is that you can be sure that this happens exactly once if you're holding the redialing mutex while setting this off. Otherwise you might get a growing number of goroutines sleeping and then refreshing the config with a corresponding unbounded increase of RPCs to Cloud SQL API backends. That'd be a fun problem to have to mitigate for sure ;)

Reply:
Moved to refreshCfg.